Dev tic tac toe

WaclaWojcieszko opened the pull request:
Wyciąłem pozostałe boczne gałęzie, bo zrobił się bałagan, a teraz przy okazji pracy domowej z czytania plików i zliczania najpospolitszych wyrazów.

WaclaWojcieszko committed:
Lesson3

WaclaWojcieszko committed:
DEV wszystko wykomentarzowane, łatwiej będzie od nowa.

WaclaWojcieszko committed:
DEV dodane interfejsy, ale nie hula.

WaclaWojcieszko committed:
DEV dodane interfejsy, ale nie hula.

WaclaWojcieszko committed:
DEV zmiany wprowadzone, ale nie kompiluje się.

WaclaWojcieszko committed:
DEV zmiany wprowadzone, kompiluje się. Brakuje czegoś klasie Man - nie mogę dotrzeć z Main do metod przynależnych Person, Student, etc.

WaclaWojcieszko committed:
DEV w ten sposób?

WaclaWojcieszko committed:
Lesson04

WaclaWojcieszko committed:
Lesson04

WaclaWojcieszko committed:
Lesson04

WaclaWojcieszko committed:
HOMEWORK04
nie jestem pewny powiązań Humana/Ai z Controllerem i jak w Main dokonywałby się wybór, jakiego typu Controller trzeba zrobić.

WaclaWojcieszko committed:
HOMEWORK04 wykomentarzowane naprawiam
nie jestem pewny powiązań Humana/Ai z Controllerem i jak w Main dokonywałby się wybór, jakiego typu Controller trzeba zrobić.

WaclaWojcieszko committed:
Lesson05

WaclaWojcieszko committed:
Homework04
Kompiluje się >>> Sukces: brak efektu!

WaclaWojcieszko committed:
DEV TicTacToe

WaclaWojcieszko committed:
DEV TicTacToe

WaclaWojcieszko committed:
Lesson06

WaclaWojcieszko committed:
Lesson07

WaclaWojcieszko committed:
Homework06

zaległa praca domowa w homework.homework06 klasa Reader

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Reader {
```
nazwa klasy Reader nie jest zbyt szczęśliwa - są już takie standardowe klasy, będzie kłopot ze zrozumieniem "co ma wspólnego ta klasa Reader z innymi klasami Reader".
Z formalnego punktu widzenia - jest ok, od tego właśnie są pakiety o unikalnych nazwach, żeby nie było problemu z kompilacją

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+
+
+            while ((s = bufferedReader.readLine()) != null) {
+                for (int n = 0; n < s.split(" ").length; ++n) {
```
Generalnie odradzam iterowanie po czymkolwiek pętlą for z licznikiem. Jeszcze będę o tym mówić w sobotę przy iteratorach.
Wczytane jest coś relatywnie prostego (String), spilt() dzieli string na tablicę String[], z czymś takim dobrze obchodzi się pętla: `for ( String word : s.split(" ") ) { .... }`
W szczególności - nie plącze Ci się po kodzie jakaś zmienna licznikowa, wszędzie tam, gdzie używasz indeksowania tablicy `sSplit[n]` używasz po prostu `word`

WaclaWojcieszko replied:
Dziękuję! Teraz chyba złapię w końcu pętlę foreach.

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+
+            while ((s = bufferedReader.readLine()) != null) {
+                for (int n = 0; n < s.split(" ").length; ++n) {
+                    String[] sSplit = s.split(" ");
```
DRY: Don't Repeat Yourself.
1. `s.split(" ")` masz już użyte w pętli, możesz się pomylić w drugim użyciu (na przykład dać dwie spacje).
2. Jak wpadniesz na pomysł, że separatorem jest na przykład dwukropek, to będziesz musiał zmienić w dwóch miejscach
3. (ja wiem, że w Javie nie dbamy nadmiernie o wydajność, ale) split w pętli jest słabo wydajne - dzielisz każdy wczytany string od nowa tyle razy, ile jest słów w stringu od nowa. tylko po to, aby wydobyć kolejne słowo.

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+                for (int n = 0; n < s.split(" ").length; ++n) {
+                    String[] sSplit = s.split(" ");
+
+                    if (!sSplit[n].equals("") && wordsMap.containsKey(sSplit[n])) {
```
Tutaj, oraz w pobliskim `else`: sprawdzenie, czy `sSplit[n].equals("")` niewiele wnosi. Pusty string `""` jest całkiem dobrym stringiem: ma swoją długość (zero), można go porównać do innego stringu, itp. Czy może chciałeś zabezpieczyć się przed nullem?

WaclaWojcieszko replied:
Otrzymywaliśmy wynik w pliku, gdzie "" było najliczniejszym wynikiem po splicie no i trochę głupio, żeby najczęstszym wyrazem był 0literowy wyraz...

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+                for (int n = 0; n < s.split(" ").length; ++n) {
+                    String[] sSplit = s.split(" ");
+
+                    if (!sSplit[n].equals("") && wordsMap.containsKey(sSplit[n])) {
```
Jeśli ani `if` ani `else` nie miało nic robić, jeśli słowo jest krótsze niż 4 znaki to można było przed nimi wszystkimi dać
```
if ( sSplit[n].length() < 4 ) { 
  continue; 
}
```

albo 

```
if ( sSplit[n].length() >= 4 ) { 
// tu całe przetwarzanie
}
```

WaclaWojcieszko replied:
Słusznie. Paulina przypomniała sobie o tym warunku jak pętla już była napisana i faktycznie nie jest to przemyślane wykonianie.

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+                        int value = wordsMap.get(sSplit[n]);
+                        System.out.println(sSplit[n]);
+                        wordsMap.replace(sSplit[n], ++value);
+                    } else if (!sSplit[n].equals("") && sSplit[n].length() > 4 && !wordsMap.containsKey(sSplit[n])) {
```
Druga część warunku `containsKey()` jest zbędna (poza tym, że DRY) - albo zwiększasz licznik, ale ustawiasz jego wartość na 1

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+                    if (!sSplit[n].equals("") && wordsMap.containsKey(sSplit[n])) {
+                        int value = wordsMap.get(sSplit[n]);
+                        System.out.println(sSplit[n]);
+                        wordsMap.replace(sSplit[n], ++value);
```
`replace` nadmiarowe, zwykłe `put` wystarczy.
O mapach wiadomo, że ich klucze są zbiorem - są unikalne. czyli takie coś:
```
wordsMap.put("Ala",1);
wordsMap.put("Ala",2);
wordsMap.put("Ala",15);
```
spowoduje powstanie jednej pozycji w mapie: "Ala", o wartości 15 (tej ostatniej)

WaclaWojcieszko replied:
W sumie do czego w takim razie ma sens użycie `replace`?

jbanaszczyk replied:
Trochę filozoficzne.
1. Jeśli w mapie nie ma jeszcze klucza, to `replace` go nie doda, a `put` tak.
2. Replace ma wersję przeciążoną (ważniejszą): `replace(K key, V oldValue, V newValue)` - podstawia `newValue` tylko jeśli istnieje klucz `key` __i__ jest zmapowany na `oldValue`. Wtdy pokazuje się prawdziwa natura `replace`: traktujemy `key oldValue` jako parę (tuple), i zastępujemy ją parą `key newValue`.

W tym konkretnym kawałku kodu - działa tak samo, kwestia czytelności - dwie linijki obok jest `put`, tu jest `replace`, dlaczego są różne???

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+                    if (!sSplit[n].equals("") && wordsMap.containsKey(sSplit[n])) {
+                        int value = wordsMap.get(sSplit[n]);
+                        System.out.println(sSplit[n]);
+                        wordsMap.replace(sSplit[n], ++value);
```
Nie ma potrzeby zwiększać value.
```
wordsMap.replace(sSplit[n], value+1);
```
W sumie cały blok można zapisać tak:
```
                        wordsMap.put(sSplit[n], wordsMap.get(sSplit[n]) +1);
```
A w wersji z moją pętlą:
```
                        wordsMap.put(word, wordsMap.get(word) +1);
```

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+
+                for (Map.Entry<String, Integer> entry : wordsMap.entrySet()) {
+
+                    if (maxEntry == null || entry.getValue().compareTo(maxEntry.getValue()) > 0) {
```
`entry.getValue()` i `maxEntrygetValue()` to są integerey, integery porównuje sie po prostu operatorem relacji:
```
entry.getValue() > maxEntry.getValue()
```

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+
+                Map.Entry<String, Integer> maxEntry = null;
+
+                for (Map.Entry<String, Integer> entry : wordsMap.entrySet()) {
```
Algorytm tej pętli wzbudził moje uznanie 👍

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+                        theWords.put(entry.getKey(), entry.getValue());
+
+
+                    } else if (maxEntry == null || entry.getValue().compareTo(maxEntry.getValue()) == 0) {
```
```
entry.getValue() == maxEntry.getValue()
```

jbanaszczyk commented on src/org/wojcieszko/homework/homework06/Reader.java:
```diff
@@ -0,0 +1,64 @@
+
+                }
+
+                Map.Entry<String, Integer> maxEntry = null;
```
Ten fragment z wyszukaniem najdłuższych słów wykonujesz od nowa w po każdym wczytaniu nowego stringu, po czym jego wyników do niczego nie używasz. Używasz tylko jeden raz - po zakończeniu czytania

jbanaszczyk reviewed (approved):
Ale się rozpisałem
Przepraszam, wczoraj nie dałem rady zajrzeć do kodu

WaclaWojcieszko committed:
Homework06

zaległa praca domowa w homework.homework06 klasa Reader

WaclaWojcieszko committed:
Lesson08

WaclaWojcieszko committed:
Lesson08

WaclaWojcieszko committed:
Homework08
Troszkę bezczelnie to zrobione może coś ciekawszego wymyślę jeszcze.

WaclaWojcieszko committed:
Homework08
Troszkę bezczelnie to zrobione może coś ciekawszego wymyślę jeszcze.
reformat

WaclaWojcieszko committed:
Homework08
Troszkę bezczelnie to zrobione może coś ciekawszego wymyślę jeszcze.

Taaaaki durny błąd w pierwszym podejściu, aż mi wstyd...

WaclaWojcieszko committed:
Lesson08

WaclaWojcieszko committed:
Merge pull request #5 from WaclaWojcieszko/Homework_second_value

Homework second value
